### PR TITLE
Kernel Bump v2

### PR DESCRIPTION
**Pull request by oliv3r**
A bunch of fixes and improvements, also based on discussions in the first version, found in #14713.

* scripts/kernel_bump: Delete merge commit
* scripts/kernel_bump: Allow bumping sub-targets
* scripts/kernel_bump: Use the git index to find the needed config files
* scripts/kernel_bump: Use git to obtain the list of files
* scripts/kernel_bump: Allow for migrating only configuration files
* scripts/kernel_bump: Do no run on dirty repositories
* scripts/kernel_bump: Drop unused function
* scripts/kernel_bump: Avoid potential copyright claim

### CODE REVIEW

**Comment by rany2** [comment]
@oliv3r Is there a reason why you aren't doing something like `git ls-files 'target/*/config-6.1'`?

**Comment by namiltd** [comment]
Maybe add a change from patches-6.1 to patches-6.6 in this version?

**Comment by oliv3r** [comment]
Nope, I think that's a much better improvement to simply rely on what got already knows.

I'll play with that and use that for my v2!

**Comment by oliv3r** [comment]
@rany2 so now it uses git ls-tree and git ls-files :D 

@namiltd what made you say that, as this was the purpose of this script. Anyway, v2 should now do what you expected?
Also, I tested x86 and seemed to work fine! I used `./scripts/kernel_bump.sh -p x86 -s 6.1 -t 6.6` to see. I've also tested with ` -r generic` which works, but then the root config file doesn't get migrated (as expected), so x86 might be a bit 'special' in that case. Though in the future, if updating only specific subtargets turns out to be a 'thing', this can be refined.

@robimarko so hopefully v2 improves, adds new features and makes things better!

@ehem this should address any concerns you have raised before! Thanks for your review, and hope to see more review here :)

**Review by rany2 — APPROVED** [review]
Looks good to me, great job!

**Comment by oliv3r** [comment]
> change while IFS=',' read -r to while read -d, -r

What would that change, other then causing:
```
^-- [SC3045](https://www.shellcheck.net/wiki/SC3045) (warning): In POSIX sh, read -d is undefined.
```

**Comment by robimarko** [comment]
@hauke Your opinion would really help here

**Comment by ehem** [comment]
I see issues to point to in a review.  I was rather hoping someone would at least _test_ #14907 before I submit a review of #14913.  Someone who cannot review is entirely capable of testing.

**Comment by ehem** [comment]
I am unsure I've quite replicated `kernel_bump.sh`'s functionality _exactly_, but I am fairly confident I've got something which is mostly compatible based on invoking `kernel_upgrade.pl` from #14907.

**Comment by namiltd** [comment]
I have three comments:
-The script does not work (err: Cannot find target linux directory.) if the name of the folder in which the branch is equal to the name of the platform
-On Windows systems (TartoiseGIT - GIT Bash here), finding the patches-_version_ folder does not work. The script only copies .config files.
-When we run the script without parameters, the following is displayed: "err: Source () and target () versions need to be defined. In my opinion, the same text as with the -h option should be displayed here.

**Comment by oliv3r** [comment]
> I have three comments: -The script does not work (err: Cannot find target linux directory.) if the name of the folder in which the branch is equal to the name of the platform 

Let me revalidate this. After the `git ls-files` change I checked that it would work (produce output) from the target directory, but as I was checking multiple platforms, I actually only tested from the root. Valueable feedback, thank you.

> -On Windows systems (TartoiseGIT - GIT Bash here), finding the patches-_version_ folder does not work. The script only copies .config files.

I don't have windows, so cannot validate this. I'd need your help to debug this. Can you set the environment variable `export DEBUG_TRACE_SH=all` and run the script again post the problematic output?

> -When we run the script without parameters, the following is displayed: "err: Source () and target () versions need to be defined. In my opinion, the same text as with the -h option should be displayed here.

Since you are the second one to say this, I will  add this (and improve the output a little). Since the idea was that if ONE parameter was missing, you get '`err: Source (v5.15) and target () versions need to be defined`. e.g. it would be obvious. I see now it is not as obvious, and will improve the message and print the usage.

**Comment by namiltd** [comment]
obraz
Folder at91 and platform at91

**Comment by robimarko** [comment]
@oliv3r The script is creating `config--v6.6` and like folders instead of `config-6.6`

**Comment by oliv3r** [comment]
> @oliv3r The script is creating `config--v6.6` and like folders instead of `config-6.6`
> 
> ```
> robimarko@fedora:~/Building/AX3600/qualcommax-6.6$ ls target/linux/qualcommax/
> config-6.1     config--v6.6   files/         image/         ipq60xx/       ipq807x/       Makefile       patches-6.1/   patches--v6.6/
> ```

Curious:

```
scripts/kernel_bump.sh -p at91 -s v5.15 -t 6.6
ls target/linux/at91/*
target/linux/at91/Makefile  target/linux/at91/modules.mk

target/linux/at91/base-files:
etc

target/linux/at91/files:
arch

target/linux/at91/image:
Config.in  dfboot  gen_at91_sdcard_img.sh  Makefile  sam9x.mk  sama5.mk  sama7.mk  uboot-env.txt

target/linux/at91/patches-5.15:
...

target/linux/at91/patches-6.6:
...

target/linux/at91/sam9x:
config-5.15  config-6.6  target.mk

target/linux/at91/sama5:
config-5.15  config-6.6  target.mk

target/linux/at91/sama7:
config-5.15  config-6.6  target.mk

```

and
```
scripts/kernel_bump.sh -p qualcommax -s v6.1 -t 6.6
ls target/linux/qualcommax/*
target/linux/qualcommax/config-6.1  target/linux/qualcommax/config-6.6  target/linux/qualcommax/Makefile

target/linux/qualcommax/files:
arch  include

target/linux/qualcommax/image:
ipq60xx.mk  ipq807x.mk  Makefile

target/linux/qualcommax/ipq60xx:
base-files  config-default  target.mk

target/linux/qualcommax/ipq807x:
base-files  config-default  target.mk

target/linux/qualcommax/patches-6.1:
...

target/linux/qualcommax/patches-6.6:
...
```

**Comment by robimarko** [comment]
> "-t 6.6" should be "-t v6.6"
> 
> But in fact the script should ignore this "v" because it is an unnecessary syntax complication.

I agree, the v is just complication

**Comment by oliv3r** [comment]
But it does. As you saw in my example I used both forms on purpose, to indicate the script ignores the leading v, so users don't have to care which format to use :)

**Comment by oliv3r** [comment]
Hah, I found the error of my ways. Or atleast, one. The leading v is stripped very early on, so early, it o ly happens for cmdline arguments, not if set via environment variables. I'll fix that!

**Comment by oliv3r** [comment]
@namiltd curious how you invoked the program however, and @robimarko that caused this weird double dash `config--v6.6` to appear. A debug lug or command invocation would be very useful.

**Comment by robimarko** [comment]
@oliv3r Well, fun fact is that today I cannot reproduce the dash bein added, it just worked as its supposed to :)

**Comment by namiltd** [comment]
@oliv3r I downloaded the branch oliv3r:kernel_bump

**Comment by oliv3r** [comment]
So the time you did this matters, as O have pushed fixes since :p

As for Robert, this could apply to you as well ;)

**Comment by robimarko** [comment]
Well, that is the thing, I did not update to your latest push before trying again and it just worked, no idea what was wrong before

**Comment by oliv3r** [comment]
> Well, that is the thing, I did not update to your latest push before trying again and it just worked, no idea what was wrong before

`history`? Maybe you invoked it with `--v` but then that doesn't make sense either ... I thought computers where quite deterministic :D

**Comment by robimarko** [comment]
> > Well, that is the thing, I did not update to your latest push before trying again and it just worked, no idea what was wrong before
> 
> `history`? Maybe you invoked it with `--v` but then that doesn't make sense either ... I thought computers where quite deterministic :D

Could be as it was quite a long day, it suddenly worked when brain was fresh in the morning

**Comment by robimarko** [comment]
BTW, @oliv3r there is now a discussion on the IRC regarding the issue that doing remove and then adding stuff later-in will break `git rebase`.

**Comment by ehem** [comment]
...and elsewhere people were complaining about frequent rebases...

> BTW, @oliv3r there is now a discussion on the IRC regarding the issue that doing remove and then adding stuff later-in will break `git rebase`.

I thought it was obvious from the start this has potentially interesting effects on `git rebase`.  The ideal rebase strategy is to use two identical branches.  Rebase one branch onto the rename commit, rebase one branch onto the end merge.  Then rebase the rename branch onto the merge branch and squash.  Suddenly your work is applied to both versions.
Not ideally convenient, but better than what was possible before.

I'm still looking for testers for #14907.  There are several issues with #14913, but since this seems the only approach for getting action on #14907...
The UI design is quite distinct and I'm unsure I've got the wrapper quite right, but I've got patches which appear to reimplement `kernel_bump.sh` as a wrapper around a single invocation of `kernel_upgrade.pl`.
```
git remote add ehem https://github.com/ehem/openwrt.git
git remote set-branches ehem bumper script
git fetch ehem
git rebase ehem/bumper ehem/script
```
Whereas `kernel_bump.sh` simply cannot replicate `kernel_upgrade.pl`'s functionality.

**Comment by ehem** [comment]
Hmm, guess I should mention.  The first of the patches for turning `kernel_bump.sh` into a wrapper for `kernel_upgrade.pl` seems suspiciously like a design suggestion.  I was created it strictly for the purpose of doing the wrapper, but does seem like it might be better for the script anyway.

**Comment by oliv3r** [comment]
What do you mean? I'm not sure I follow ...

**Comment by oliv3r** [comment]
As discussed on IRC, breaking got bisect is expect but shouldn't be a problem. I'm campaigning on getting something into fit natively, but having a wrapper script for bisect, and using `git notes ` is certainly feasible. Without native git support, it does mean bisect requires a wrapper, which is unfortunate of course but hardly a problem. The discussion then becomes, use a script vs loosing history (ish) which I think is to e far more often.

**Comment by ehem** [comment]
Your script is doing most processing of variables (ensuring they're set, adding/removing prefixes, etc) in your `main()` function.  Yet `$platform_name` is left mostly unprocessed by `main()` and instead adjusted inside `bump_kernel()`.

For the conversion to wrapper I had to fix this.  I suspect you might want to consider grabbing the first commit.

**Comment by oliv3r** [comment]
> Your script is doing most processing of variables (ensuring they're set, adding/removing prefixes, etc) in your `main()` function. Yet `$platform_name` is left mostly unprocessed by `main()` and instead adjusted inside `bump_kernel()`.
> 
> For the conversion to wrapper I had to fix this. I suspect you might want to consider grabbing the first commit.

The main function does just some argument setup, it's part of the arguments. You could argue it should be part of init maybe, but not main.

File: scripts/kernel_bump.sh [hunk 1]
```diff
@@ -74,67 +79,89 @@ cleanup()
...
-	for _path in "${_target_dir}/"*; do
-		if [ ! -s "${_path}" ] || \
-		   [ "${_path}" = "${_path%%"-${source_version}"}" ]; then
+	if [ "${config_only:-false}" != 'true' ]; then
+		for _path in $(git ls-tree -d -r --name-only '__openwrt_kernel_files_mover' "${_target_dir}" |
+			       sed -n "s|^\(.*-${source_version}\).*|\1|p" |
+			       sort -u); do
+			if [ ! -s "${_path}" ] || \
+			   [ "${_path}" = "${_path%%"-${source_version}"}" ]; then
+				continue
```
Comment (namiltd) [inline scripts/kernel_bump.sh:126]: This testing doesn't work for Windows systems (TartoiseGIT - GIT Bash here). For Windows, folders always have size 0.

**Comment by oliv3r** [comment]
Gah, OK I'll change that. I prefer normally using g -s as it works on files, symlinks and directories in Linux on my fs at least. The difference compared to -e is that it will fail on a zero byte file.

Thanks for debugging stupid NTFS.

**Comment by ehem** [comment]
> > Your script is doing most processing of variables (ensuring they're set, adding/removing prefixes, etc) in your `main()` function. Yet `$platform_name` is left mostly unprocessed by `main()` and instead adjusted inside `bump_kernel()`.
> > For the conversion to wrapper I had to fix this. I suspect you might want to consider grabbing the first commit.
> 
> The main function does just some argument setup, it's part of the arguments. You could argue it should be part of init maybe, but not main.

The section at the top of `bump_kernel()` looks suspiciously like argument setup.

I'll admit I'm more concerned about showing `kernel_upgrade.pl` can do everything this does.  This isn't something I could reject as a reviewer (something about ethics and conflict of interest), but it does seem a bit odd.

Meanwhile there are some rather major issues waiting in a review submission, but I'm awaiting at least testing on another pull...

**Comment by ehem** [comment]
Uhm...  No activity?

**Comment by oliv3r** [comment]
There is, just behind the shadows :)

**Comment by ehem** [comment]
I do have a review pretty well ready, but my theory was this was in exchange for assistance elsewhere (even if you cannot review, having someone simply state other things _work_ has value).

**Comment by oliv3r** [comment]
I remember, but super busy with $work

**Comment by robimarko** [comment]
Thanks! Rebased on top of main and merged!

**Comment by ehem** [comment]
> I remember, but super busy with $work

Still listening to the crickets here.

**Pull request merged**